[JSC] DestructuringAssignmentTarget should be evaluated prior to calling [[Get]] / stepping iterator

shvaikalesh opened the pull request:
#### e14851d085d342b16e9b2ed95abaf77957d21f06
<pre>
[JSC] DestructuringAssignmentTarget should be evaluated prior to calling [[Get]] / stepping iterator
<a href="https://bugs.webkit.org/show_bug.cgi?id=268272">https://bugs.webkit.org/show_bug.cgi?id=268272</a>
&lt;<a href="https://rdar.apple.com/problem/121960976">rdar://problem/121960976</a>&gt;

Reviewed by Keith Miller and Yusuke Suzuki.

ECMA-262 specifies left-to-right evaluation order of expressions, which is maintained by evaluating
a DestructuringAssignmentTarget (unless it is a destructuring pattern) prior to calling [[Get]] [1] or
stepping the iterator [2] or evaluating the Initializer. The same applies [3][4] to a ...rest parameter.

This patch makes a copy of AssignmentElementNode::bindValue() for cases except ResolveNode and
divides it into emitNodesForDestructuring() / bindValueWithEmittedNodes().
A few implementation notes:
  a) AssignmentElementNode is guaranteed not to be a destructuring pattern;
  b) when binding to a ResolveNode, all errors occur in PutValue [5] rather then during evaluation
     of DestructuringAssignmentTarget;
  c) in order to prevent register clash when emitting an op_call to @copyDataProperties(),
     emitNodesForDestructuring() accepts temporary registers created before CallArguments.

To achieve complete spec compatibility, emitToPropertyKeyOrNumber() is called for computed property
names before evaluating DestructuringAssignmentTarget [6] regardless of ...rest parameter presence,
and is skipped only for commonly known types (string, number) to preserve performance.

Aligns JSC with the spec, V8, and SpiderMonkey. Proven to be perf-neutral on JS2 and SP3.

[1]: <a href="https://tc39.es/ecma262/#sec-runtime-semantics-keyeddestructuringassignmentevaluation">https://tc39.es/ecma262/#sec-runtime-semantics-keyeddestructuringassignmentevaluation</a> (step 1)
[2]: <a href="https://tc39.es/ecma262/#sec-runtime-semantics-iteratordestructuringassignmentevaluation">https://tc39.es/ecma262/#sec-runtime-semantics-iteratordestructuringassignmentevaluation</a> (production before last, step 1)
[3]: <a href="https://tc39.es/ecma262/#sec-runtime-semantics-restdestructuringassignmentevaluation">https://tc39.es/ecma262/#sec-runtime-semantics-restdestructuringassignmentevaluation</a> (step 1)
[4]: <a href="https://tc39.es/ecma262/#sec-runtime-semantics-iteratordestructuringassignmentevaluation">https://tc39.es/ecma262/#sec-runtime-semantics-iteratordestructuringassignmentevaluation</a> (last production, step 1)
[5]: <a href="https://tc39.es/ecma262/#sec-putvalue">https://tc39.es/ecma262/#sec-putvalue</a>
[6]: <a href="https://tc39.es/ecma262/#sec-runtime-semantics-propertydestructuringassignmentevaluation">https://tc39.es/ecma262/#sec-runtime-semantics-propertydestructuringassignmentevaluation</a> (last production, step 1)

* JSTests/stress/destructuring-evaluation-order-array-bracket-accessor-iterator-done.js: Added.
* JSTests/stress/destructuring-evaluation-order-array-bracket-accessor.js: Added.
* JSTests/stress/destructuring-evaluation-order-array-dot-accessor-iterator-done.js: Added.
* JSTests/stress/destructuring-evaluation-order-array-dot-accessor.js: Added.
* JSTests/stress/destructuring-evaluation-order-array-rest-element-iterator-done.js: Added.
* JSTests/stress/destructuring-evaluation-order-array-rest-element.js: Added.
* JSTests/stress/destructuring-evaluation-order-object-bracket-accessor.js: Added.
* JSTests/stress/destructuring-evaluation-order-object-dot-accessor.js: Added.
* JSTests/stress/destructuring-evaluation-order-object-rest-element.js: Added.
* JSTests/test262/expectations.yaml: Mark 5 tests as passing, progress 56 tests.
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::ArrayPatternNode::bindValue const):
(JSC::ObjectPatternNode::bindValue const):
(JSC::AssignmentElementNode::emitNodesForDestructuring const):
(JSC::AssignmentElementNode::bindValueWithEmittedNodes const):
* Source/JavaScriptCore/parser/Nodes.h:

Canonical link: <a href="https://commits.webkit.org/281013@main">https://commits.webkit.org/281013@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/31219f3bdf1478ffec81885805c143c867b7d02f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58427 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37754 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10911 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62052 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/8871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45390 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9068 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/47321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/8871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60458 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/35363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/50518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/28165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/32118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/7817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7875 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/51518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/54071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/8087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63756 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/57669 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2340 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/8119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/54642 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2348 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/50540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/54708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/1980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/79429 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8702 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/33583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/13214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/34669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/35753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/34414 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (commented):
Nice

Constellation reviewed (commented):
Did you measure SP3 / JS2 results with this change?

kmiller68 commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -5527,6 +5527,10 @@ void ArrayPatternNode::bindValue(BytecodeGenerator& generator, RegisterID* rhs)
 
     RefPtr<RegisterID> done;
     for (auto& target : m_targetPatterns) {
+        std::optional<BaseAndPropertyName> targetBaseAndPropertyName = std::nullopt;
```
Nit: you don't need the `= std::nullopt` since that's the default value for `std::optional`

shvaikalesh replied:
Removed all the `std::nullopt`, thanks!!

kmiller68 commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -5707,7 +5722,10 @@ void ObjectPatternNode::bindValue(BytecodeGenerator& generator, RegisterID* rhs)
                 else
                     temp = generator.newTemporary();
 
+                std::optional<BaseAndPropertyName> targetBaseAndPropertyName = std::nullopt;
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -5719,28 +5737,38 @@ void ObjectPatternNode::bindValue(BytecodeGenerator& generator, RegisterID* rhs)
                 ASSERT(target.bindingType == BindingType::RestElement);
                 ASSERT(i == m_targetPatterns.size() - 1);
 
+                std::optional<BaseAndPropertyName> targetBaseAndPropertyName = std::nullopt;
```
Ditto.

Constellation reviewed (approved):
r=me

kmiller68 reviewed (approved):
r=me with nits.

Commit:
[JSC] DestructuringAssignmentTarget should be evaluated prior to calling [[Get]] / stepping iterator
https://bugs.webkit.org/show_bug.cgi?id=268272
<rdar://problem/121960976>

Reviewed by Keith Miller and Yusuke Suzuki.

ECMA-262 specifies left-to-right evaluation order of expressions, which is maintained by evaluating
a DestructuringAssignmentTarget (unless it is a destructuring pattern) prior to calling [[Get]] [1] or
stepping the iterator [2] or evaluating the Initializer. The same applies [3][4] to a ...rest parameter.

This patch makes a copy of AssignmentElementNode::bindValue() for cases except ResolveNode and
divides it into emitNodesForDestructuring() / bindValueWithEmittedNodes().
A few implementation notes:
  a) AssignmentElementNode is guaranteed not to be a destructuring pattern;
  b) when binding to a ResolveNode, all errors occur in PutValue [5] rather then during evaluation
     of DestructuringAssignmentTarget;
  c) in order to prevent register clash when emitting an op_call to @copyDataProperties(),
     emitNodesForDestructuring() accepts temporary registers created before CallArguments.

To achieve complete spec compatibility, emitToPropertyKeyOrNumber() is called for computed property
names before evaluating DestructuringAssignmentTarget [6] regardless of ...rest parameter presence,
and is skipped only for commonly known types (string, number) to preserve performance.

Aligns JSC with the spec, V8, and SpiderMonkey. Proven to be perf-neutral on JS2 and SP3.

[1]: https://tc39.es/ecma262/#sec-runtime-semantics-keyeddestructuringassignmentevaluation (step 1)
[2]: https://tc39.es/ecma262/#sec-runtime-semantics-iteratordestructuringassignmentevaluation (production before last, step 1)
[3]: https://tc39.es/ecma262/#sec-runtime-semantics-restdestructuringassignmentevaluation (step 1)
[4]: https://tc39.es/ecma262/#sec-runtime-semantics-iteratordestructuringassignmentevaluation (last production, step 1)
[5]: https://tc39.es/ecma262/#sec-putvalue
[6]: https://tc39.es/ecma262/#sec-runtime-semantics-propertydestructuringassignmentevaluation (last production, step 1)

* JSTests/stress/destructuring-evaluation-order-array-bracket-accessor-iterator-done.js: Added.
* JSTests/stress/destructuring-evaluation-order-array-bracket-accessor.js: Added.
* JSTests/stress/destructuring-evaluation-order-array-dot-accessor-iterator-done.js: Added.
* JSTests/stress/destructuring-evaluation-order-array-dot-accessor.js: Added.
* JSTests/stress/destructuring-evaluation-order-array-rest-element-iterator-done.js: Added.
* JSTests/stress/destructuring-evaluation-order-array-rest-element.js: Added.
* JSTests/stress/destructuring-evaluation-order-object-bracket-accessor.js: Added.
* JSTests/stress/destructuring-evaluation-order-object-dot-accessor.js: Added.
* JSTests/stress/destructuring-evaluation-order-object-rest-element.js: Added.
* JSTests/test262/expectations.yaml: Mark 5 tests as passing, progress 56 tests.
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::ArrayPatternNode::bindValue const):
(JSC::ObjectPatternNode::bindValue const):
(JSC::AssignmentElementNode::emitNodesForDestructuring const):
(JSC::AssignmentElementNode::bindValueWithEmittedNodes const):
* Source/JavaScriptCore/parser/Nodes.h:

Canonical link: https://commits.webkit.org/281013@main
